### PR TITLE
Switch from callbacks to promises for findRoad

### PR DESCRIPTION
**Pull request by JakeChampion**
Callbacks are useful when an async action may fire multiple times, promises are useful when an async action only fires once. They also benefit by not inverting the control, which is what callbacks do.

### CODE REVIEW

**Comment by JakeChampion** [comment]
This has changed the signature of `findRoad`, the consumer on this function would have to be updated to consume promises instead of passing a callback.

**Comment by aksiksi** [comment]
Thanks for the changes! I've merged them into the `dev` branch, but I think I'm handling the returned Promise from `findRoad` incorrectly. What's confusing is that the first Promise can return either a response object or another Promise. How can that be handled?

File: server/backend/server.js [hunk 1]
```diff
@@ -52,19 +54,15 @@ var decrypt = (req) => {
   1. Queries `segments` collection to find the road the coordinate lies on.
   2. If matched, queries `roads` collection to determine road name.
 
-  Returns a Response object.
+  Returns a Promise wrapped Response object.
 */
-var findRoad = (parsed, db, callback) => {
-  const resp = newResponse();
-
+const findRoad = ({lat, lng}, db) => {
```
Comment (JakeChampion) [inline server/backend/server.js:59]: `findRoad` should only be returning a Promise wrapped Response object. I'll comment on the lines with explanations.

File: server/backend/server.js [hunk 2]
```diff
@@ -75,34 +73,33 @@ var findRoad = (parsed, db, callback) => {
...
-      // No segment match!
-      callback(resp);
+  const resp = newResponse();
+  const segments = db.collection(config.segments);
+  const findOneSegment = denodeifyMethod(segments, 'findOne');
+
+  return findOneSegment(q, {speed: 1, road_id: 1, _id: 0})
+  .then({road_id, speed} => {
+    if (!segment) {
+      return resp;
```
Comment (JakeChampion) [inline server/backend/server.js:83]: Returning anything from a `thenable` will wrap it in a `Promise`, this enables chaining and error propagation to the consumer.

This return statement is called when `findOneSegment` executes successfully but does not find a segment.

File: server/backend/server.js [hunk 3]
```diff
@@ -75,34 +73,33 @@ var findRoad = (parsed, db, callback) => {
...
+      })
+      .catch(err => {
+        resp.name = "";
+        return resp;
+      }
     }
+  })
+  .catch(err => {
+    resp.online = 0;
+    return resp;
```
Comment (JakeChampion) [inline server/backend/server.js:102]: This return statement is called when `findOneSegment` executes unsuccessfully.

File: server/backend/server.js [hunk 4]
```diff
@@ -75,34 +73,33 @@ var findRoad = (parsed, db, callback) => {
...
-          resp.found = 1;
-          resp.speed = segment.speed;
-          resp.name = road.name;
-        }
-
-        callback(resp);
-      });
+      const roads = db.collection(config.roads);
+      const findOneRoad = denodeifyMethod(roads, 'findOne');
+      return findOneRoad({_id: road_id})
```
Comment (JakeChampion) [inline server/backend/server.js:87]: This is returning a `Promise`, however the promise already has some `thenables` chained onto the end. The return of last `thenable` or the return of the `catch` (if a `catch` is attached) is what gets returned to the consumer of this `findRoad`.

File: server/backend/server.js [hunk 5]
```diff
@@ -75,34 +73,33 @@ var findRoad = (parsed, db, callback) => {
...
+      .then(road => {
+        resp.found = 1;
+        resp.speed = speed;
+        resp.name = road.name;
+        return resp;
```
Comment (JakeChampion) [inline server/backend/server.js:92]: This return statement is reached if a segment and a road are found. The Response object `resp` is returned to the consumer of `findRoad`.

File: server/backend/server.js [hunk 6]
```diff
@@ -75,34 +73,33 @@ var findRoad = (parsed, db, callback) => {
...
+      })
+      .catch(err => {
+        resp.name = "";
+        return resp;
```
Comment (JakeChampion) [inline server/backend/server.js:96]: This return statement is reached if a segment is found but an error was thrown when trying to find a road. The Response object `resp` is returned to the consumer of `findRoad`.

**Comment by JakeChampion** [comment]
If you haven't used Promises a lot, Nolan Lawson has a written quite a good blog post on the topic, it includes a short quiz :-) - http://pouchdb.com/2015/05/18/we-have-a-problem-with-promises.html

**Comment by aksiksi** [comment]
It seems there was a slight error which I fixed in the latest commit. Thanks for your contribution!

By the way, you are not appearing in the Contributors graph because you haven't added the email you used to commit to Github.

**Pull request merged**